lib/Dialect/TTIR/Transforms/Transforms.cpp

```diff
@@ -151,4 +162,317 @@ class TTIRSlidingWindow2dFixShapes
+      ConvolutionDimension::FEATURE;
+
+  int64_t spatial_count = 0;
+  for (int64_t spatial_dim : op.getConvLayout().getInputSpatialDimensions()) {
+    input_layout[spatial_dim] = spatial_count;
+    spatial_count++;
+  }
+
+  const std::vector<int64_t> desired_input_layout = ttnn_convolution_layout;
+  for (int64_t i = 0; i < static_cast<int64_t>(input_layout.size()); i++) {
+    if (input_layout[i] != desired_input_layout[i]) {
+      int64_t dim0 = i;
+      int64_t dim1 = std::find(input_layout.begin(), input_layout.end(),
+                               desired_input_layout[i]) -
+                     input_layout.begin();
+      transpose_indices.push_back(std::make_tuple(dim0, dim1));
+      std::swap(input_layout[dim0], input_layout[dim1]);
+    }
+  }
+
+  return transpose_indices;
+}
+
+/**
+ * Although this function is mostly a clone of generateInputTransposeIndices,
+ * its slightly different in that if the original Convolution op had the same
+ * input and output layout, this function will generate the same transposes,
+ * that were applied to the input but in reverse order. This makes optimizing
+ * away the inserted transposes easier.
+ */
+static std::vector<TransposeDims> generateOutputTransposeIndices(
+    ConvolutionOp op, const std::vector<int64_t> ttnn_convolution_layout) {
+  std::vector<TransposeDims> transpose_indices;
+
+  std::vector<int64_t> desired_output_layout(ttnn_convolution_layout.size(),
+                                             ConvolutionDimension::INVALID_DIM);
+  desired_output_layout[op.getConvLayout().getOutputBatchDimension()] =
+      ConvolutionDimension::BATCH;
+  desired_output_layout[op.getConvLayout().getOutputFeatureDimension()] =
+      ConvolutionDimension::FEATURE;
+
+  int64_t spatial_count = 0;
+  for (int64_t spatial_dim : op.getConvLayout().getOutputSpatialDimensions()) {
+    desired_output_layout[spatial_dim] = spatial_count;
+    spatial_count++;
+  }
+
+  std::vector<int64_t> output_layout = ttnn_convolution_layout;
+
+  for (int64_t i = static_cast<int64_t>(desired_output_layout.size()) - 1;
+       i >= 0; i--) {
+    if (desired_output_layout[i] != output_layout[i]) {
+      int64_t dim0 = i;
+      int64_t dim1 = std::find(output_layout.begin(), output_layout.end(),
+                               desired_output_layout[i]) -
+                     output_layout.begin();
+      transpose_indices.push_back(std::make_tuple(dim0, dim1));
+      std::swap(output_layout[dim0], output_layout[dim1]);
+    }
+  }
+
+  return transpose_indices;
+}
+
+static Value
+generateTransposeSequence(Value input, PatternRewriter &rewriter,
+                          std::vector<TransposeDims> transpose_indices,
+                          ::mlir::ArrayAttr operandConstraints) {
+  for (auto [dim0, dim1] : transpose_indices) {
+    input = generateTranspose(input, dim0, dim1, rewriter, operandConstraints)
+                .getResult();
+  }
+
+  return input;
+}
```

Comment:
Should all of the above be inside of `ConvolutionToConv2dPatternRewriter`? Or do you think these free functions / types will be used by other passes?

Reply:
I wrote these so that they could be used to permute inputs/weights for a convolution of any dimensionality, so I put them outside on purpose.